google/cloud/testing_util/validate_metadata.cc

```diff
@@ -87,14 +87,10 @@ RoutingHeaders ExtractMDFromHeader(std::string header) {
  * `GTEST_USES_POSIX_RE`.
  */
 MATCHER_P(MatchesGlob, glob, "matches the glob: \"" + glob + "\"") {
-  // Translate the glob into a regex pattern.
-  auto matcher1 = absl::StrReplaceAll(glob, {{"*", "[^/]+"}});
-  // Create a second matcher that replaces all "/" with "%2F" to match the
+  // Create a matcher that replaces all "/" with "%2F" to match the
   // character replacement in `internal::UrlEncode`.
-  auto decoded_glob = absl::StrReplaceAll(glob, {{"/", "%2F"}});
-  auto matcher2 = absl::StrReplaceAll(decoded_glob, {{"*", "[^/]+"}});
-  auto regex_string = absl::StrCat(matcher1, "|", matcher2);
-  std::regex regex(regex_string);
+  auto matcher = absl::StrReplaceAll(glob, {{"/", "%2F"}, {"*", "[^/]+"}});
```

Comment:
The second replacement is still strange. It is really a `{"*", ".+"}`, because we have already removed all of the `/`s.

Reply:
Note that the implementation is not perfect. e.g. Let's say the pre-encoding metadata is "projects/foo/instances/bar", but the glob is "projects/*".

```
EXPECT_FALSE("projects%2Ffoo%2Finstances%2Fbar", MatchesGlob("projects/*"))
      << "shouldn't match because `*` goes until it hits a `/`";
```

`MatchesGlob()` would return true, but we wouldn't want it to.

Reply:
I better tactic might be to "UrlDecode(arg)" before matching, rather than trying to "UrlEncode(glob)".

Reply:
I think the UrlDecode idea is good. Changed.